### ISSUE TITLE
Return statement prevents -Wsometimes-uninitialized warning

### ISSUE THREAD

**Issue by sfc-gh-sgiesecke** [llvm/llvm-project#60409]
As expected, clang produces a -Wsometimes-uninitialized warning with the following code:
```c++
int foo(bool flag, bool otherFlag) {

    int bar;

    if (flag) {
        bar = 1;
    }
    if (otherFlag) {
        // return 1;
    }

    return bar;
}
```
saying
```console
<source>:5:9: error: variable 'bar' is used uninitialized whenever 'if' condition is false [-Werror,-Wsometimes-uninitialized]
    if (flag) {
        ^~~~
<source>:12:12: note: uninitialized use occurs here
    return bar;
           ^~~
<source>:5:5: note: remove the 'if' if its condition is always true
    if (flag) {
    ^~~~~~~~~~
<source>:3:12: note: initialize the variable 'bar' to silence this warning
    int bar;
           ^
            = 0
```

However, if I uncomment the `return` statement, no warning is produced:
```c++
int foo(bool flag, bool otherFlag) {

    int bar;

    if (flag) {
        bar = 1;
    }
    if (otherFlag) {
        return 1;
    }

    return bar;
}
```

However, the situation about the potential read of the uninitialized `bar` variable doesn't change. Is this a bug in the `-Wsometimes-uninitialized` implementation or intentional?

**Comment by AaronBallman** [comment]
> However, the situation about the potential read of the uninitialized bar variable doesn't change. Is this a bug in the -Wsometimes-uninitialized implementation or intentional?

It's intentional -- we cover that situation under a different diagnostic flag so users have more control over chattiness: https://godbolt.org/z/P8YxvWx7e

It is a bit interesting that this diagnostic is not under `-Wuninitalized` though (that may still be on purpose due to false positive rates).

Closing as not a bug.

**Comment by sfc-gh-sgiesecke** [comment]
Thanks! I wasn't aware of `-Wconditional-uninitialized`. I checked it out, and it indeed has a rather high false-positive rate (where the _access_ of the uninitialized variable is dependent on a conditional expression, e.g. some "isValid" flag), which is unfortunate. Just from a users perspective (I don't know how this is implemented), the difference between `-Wsometimes-uninitialized` and `-Wconditional-uninitialized` is somewhat hard to grasp, and the documentation doesn't say anything about it: https://clang.llvm.org/docs/DiagnosticsReference.html#wconditional-uninitialized resp. https://clang.llvm.org/docs/DiagnosticsReference.html#wsometimes-uninitialized. The situation caught by `-Wsometimes-uninitialized` also _involves_ a conditional, the difference between the two situations is the `return` (which is not dependent on the possibly-uninitialized variable). I wonder if it were possible to have a warning that catches _this_ situation without incurring the high false-positive rate of `-Wconditional-uninitialized` and has the better diagnostic messages (explaining the conditions) of `-Wsometimes-uninitialized`.

**Comment by sfc-gh-sgiesecke** [comment]
Re false positives of `-Wconditional-uninitialized`, it typically catches situations like this:
```
// bool bar(int&);

int foo() {
  int outValue;
  if (!bar(outValue)) {
    return -1;
  }
  return outValue;
}
```

These seem to be very different from the one I reported here, again from a user's perspective. In particular, the second condition in my example (otherFlag) doesn't involve the uninitialized variable at all, as opposed to this one, so it can never be initialized by that expression. IIUC, the fact that the compiler cannot tell when an expression like `bar(outValue)` initializes `outValue` is the cause of the high false-positive rate.
